Initial version of unmanaged IPv6 (#116)
theforeman/foreman-documentation: guides/doc-Provisioning_Guide/topics/proc_installing_ipv6_environment.adoc

= Installing IPv6 environment

{Project} allows provisioning of {RHEL} 8 or newer over IPv6 provisioning networks. Although provisioning templates which ship with {Project} include IPv6 support for PXE and HTTP (iPXE) scenarios, the only tested and certified workflow is UEFI HTTP Boot with unmanaged DHCPv6 server which is described in this chapter.

== Limitations

> DHCP server in {RHEL} (ISC DHCP) does not provide an integration API for managing IPv6 records, therefore {SmartProxy} DHCP plugin that would provide DHCP management of IPv6 subnets does not exist. In this scenario, DHCP IPv6 must be deployed as a separate (unmanaged) service to boostrap client nodes into grub2 which then configures IPv6 networking either using DHCPv6 or via static IPv6 address.

Reviewer: Afaik grub is an abbreviation so we should probably use `GRUB`

Author: Leaving on the docs team.

== Requirements

ifeval::["{build}" == "satellite"]
{SmartProxy} must be installed on {RHEL} version 7.9 or higher.
+

> {Project} or {SmartProxy} can be only installed on IPv6-only systems, dualstack is known to work but not supported.

Reviewer: Here you use `dualstack`, later in the document you use `dual-stack`. Not sure which one is right, but we should use only one of the forms

endif::[]

ifeval::["{build}" != "satellite"]
{SmartProxy} must be installed on a system with grub2 version 2.05 or higher or system with fixes for HTTP Boot, for example CentOS version 7.9 or 8.3 or higher. For other operating systems, copy newest grub build to `/var/lib/tftpboot/grub2/grubx64.efi`.
endif::[]

> The workflow requires {Project} deoloyment with IPv6 only networking, dual-stack setup has not been tested. In the examples below, IPv6 subnet `2001:db8::/64` is subject for provisioning and {ProjectServer} has IPv6 address `2001:db8::2`.

Reviewer: `s/deoloyment/deployment`

[options="nowrap" subs="+quotes,attributes"]
----
# nmcli c add con-name eth0 type ethernet ifname eth0 \
        ip6 2001:db8::2/64 gw6 2001:db8::1 \
        ipv6.dns 2001:db8::1
----

Since clients which are booting from network cannot be identified by remote IPv6 address due to unmanaged DHCPv6 service, provisioning tokens must be enabled. To check that, go to Administer - Settings - Provisioning and check that Token duration is not set to zero.

It is assumed, native IPv6 internet connectivity is availble. If not, it's recommended to configure a HTTP proxy in order to download RPM packages:

ifeval::["{build}" == "satellite"]
[options="nowrap" subs="+quotes,attributes"]
----
# cat /etc/rhsm/rhsm.conf
proxy_hostname = myproxy.example.com
proxy_port = 8080
proxy_user = optional_proxy_username
proxy_password = optional_proxy_password
----
endif::[]

ifeval::["{build}" != "satellite"]
----

> echo "proxy=http://[2001:db8::::1]:8888" >> /etc/yum.conf

Reviewer 2: Is this a weird yum thing that it needs all semi colons or just a typo?
```suggestion
echo "proxy=http://[2001:db8::1]:8888" >> /etc/yum.conf
```

----
endif::[]

== Configuring {Project} for UEFI HTTP Boot IPv6 provisioning

To configure {Project} for UEFI HTTP Boot IPv6 provisioning, {SmartProxy} main HTTP endpoint must be enabled (the default port is 8000), TFTP and HTTP boot modules must be enabled as well. Discovery and DNS features can be optionally enabled however DHCP management must be either disabled in the installer or not used meaning that all IPv6 Subnets created in {Project} interface must have DHCP {SmartProxy} set to blank.

[options="nowrap" subs="+quotes,attributes"]
----
foreman-installer \
 --foreman-proxy-http true \
 --foreman-proxy-httpboot true \
 --foreman-proxy-tftp true \
 --foreman-proxy-dhcp false \
 --foreman-proxy-dns true \
 --foreman-proxy-dns-zone "example.com" \
 --foreman-proxy-dns-reverse "c.c.0.0.b.b.b.b.a.a.a.a.0.0.d.f.ip6.arpa"
 ----

NOTE: The following paragraph can be a module "How to set up HTTP proxy in {Project}" that can be reused across documents.

> If native IPv6 connectivity is not avilable, define HTTP proxy in {Project} and set it as the default proxy in settings.

Reviewer: `s/avilable/available`

----
hammer http-proxy create --name "Default proxy" --url http://myproxy.example.com:8888 --organizations "Default Organization" --locations "Default Location"
hammer settings set --name content_default_http_proxy --value "Default proxy"
----

NOTE: The following can be a module/procedure. We need to drop a note that this is only relevant when Katello is installed if upstream.

Import a manifest and synchronize {RHEL} 8 kickstart repository.

== Creating new IPv6 Subnet

To create new IPv6 Subnet, go to Infrastructure - Subnets and create new entry:

Protocol: IPv6
Network Address: valid IPv6 address
Network Prefix: valid prefix
Primary DNS Server: must be valid IPv6 address of a DNS server

IPAM modes:

EUI-64 - IPv6 address is calculated from host provisioning interface MAC address
Internal DB - IPv6 address is selected from a sequence of all available addresses from the subnet
None - IPv6 address must be entered manually

Boot mode can be set both to DHCP or static, keep in mind that external DHCPv6 server is deployed and when this option is set to static a DHCPv6 service must be deployed on the network for the initial HTTP UEFI bootstrap.

> NOTE: Primary DNS server must be present becuase some older versions of {RHEL} 8 do not correctly interper DNS server coming from DHCPv6 response.

Reviewer 2: ```suggestion
NOTE: Primary DNS server must be present because some older versions of {RHEL} 8 do not correctly interpret a DNS server coming from DHCPv6 response.
```

To create a IPv6 subnet via hammer do:

[options="nowrap" subs="+quotes,attributes"]
----
# hammer subnet create --name ipv6_subnet --network-type IPv6 --network 2001:db8:: --prefix 64 --tftp-id 1 --discovery-id 1 --httpboot-id 1 --template-id 1 --domains example.com
----

== Configuring external DHCPv6

> Both DHCPv6 and Router Advertisment services must be running on all provisioning networks. It can be installed on any host in the network, or on DHCP {SmartProxy}. {Project} currently does not manage DHCPv6 records, it is only used to bootstrap clients into installer.

Reviewer: `s/Advertisment/Advertisement`

= Installing IPv6 environment

{Project} allows provisioning of {RHEL} 8 or newer over IPv6 provisioning networks. Although provisioning templates which ship with {Project} include IPv6 support for PXE and HTTP (iPXE) scenarios, the only tested and certified workflow is UEFI HTTP Boot with unmanaged DHCPv6 server which is described in this chapter.

== Limitations

DHCP server in {RHEL} (ISC DHCP) does not provide an integration API for managing IPv6 records, therefore {SmartProxy} DHCP plugin that would provide DHCP management of IPv6 subnets does not exist. In this scenario, DHCP IPv6 must be deployed as a separate (unmanaged) service to boostrap client nodes into grub2 which then configures IPv6 networking either using DHCPv6 or via static IPv6 address.

== Requirements

> The workflow requires {Project} deoloyment with IPv6 only networking, dual-stack setup has not been tested. In the examples below, IPv6 subnet `2001:db8::/64` is subject for provisioning and {ProjectServer} has IPv6 address `2001:db8::2`.

Reviewer: `s/deoloyment/deployment/`

Author: Thanks.

= Installing IPv6 environment

> {Project} allows provisioning of {RHEL} 8 or newer over IPv6 provisioning networks. Although provisioning templates which ship with {Project} include IPv6 support for PXE and HTTP (iPXE) scenarios, the only tested and certified workflow is UEFI HTTP Boot with unmanaged DHCPv6 server which is described in this chapter.

Reviewer 2: ```suggestion
{Project} allows provisioning of {RHEL} 8 or newer over IPv6 provisioning networks. Although provisioning templates which ship with {Project} include IPv6 support for PXE and HTTP (iPXE) scenarios, the only tested and certified workflow is UEFI HTTP Boot with an unmanaged DHCPv6 server which is described in this chapter.
```

Author: Fixed.

== Limitations

> DHCP server in {RHEL} (ISC DHCP) does not provide an integration API for managing IPv6 records, therefore {SmartProxy} DHCP plugin that would provide DHCP management of IPv6 subnets does not exist. In this scenario, DHCP IPv6 must be deployed as a separate (unmanaged) service to boostrap client nodes into grub2 which then configures IPv6 networking either using DHCPv6 or via static IPv6 address.

Reviewer 2: ```suggestion
DHCP server in {RHEL} (ISC DHCP) does not provide an integration API for managing IPv6 records, therefore the {SmartProxy} DHCP plugin that would provide DHCP management of IPv6 subnets does not exist. In this scenario, DHCP IPv6 must be deployed as a separate (unmanaged) service to bootstrap client nodes into grub2 which then configures IPv6 networking either using DHCPv6 or via static IPv6 addresses.
```

== Requirements

ifeval::["{build}" == "satellite"]
{SmartProxy} must be installed on {RHEL} version 7.9 or higher.
+

> {Project} or {SmartProxy} can be only installed on IPv6-only systems, dualstack is not supported.

Reviewer 3: This sounds like a crucial thing, let me follow up with questions:
If users want to manage IPv6 networks, they must keep in mind that they can only do this if they manage ONLY IPv6 networks on Both {Project} and {SmartProxy}?
Is that possible to run {Project{ on an IPv4 network and {SmartProxy} on an IPv6 network? Or the whole setup including all proxies must be IPv6 only?

Author: Yes, we know this is bold limitation. We want to start with this for 6.8 and see if we can relax this in the future release. We can perhaps change the wording since the message we want to say is that we only test IPv6 setup on singlestack IPv6 installations where both Server and Proxy are IPv6 only (no dualstack). We can even relax this today with it is known to work, however we haven't verified all the features. Let's see if you can find something out of this :-)

Author: I am adding "known to work" note. This is an approved term by Satellite PMs.

[options="nowrap" subs="+quotes,attributes"]
----
# nmcli c add con-name eth0 type ethernet ifname eth0 \
        ip6 2001:db8::2/64 gw6 2001:db8::1 \
        ipv6.dns 2001:db8::1
----

Since clients which are booting from network cannot be identified by remote IPv6 address due to unmanaged DHCPv6 service, provisioning tokens must be enabled. To check that, go to Administer - Settings - Provisioning and check that Token duration is not set to zero.

It is assumed, native IPv6 internet connectivity is availble. If not, it's recommended to configure a HTTP proxy in order to download RPM packages:

ifeval::["{build}" == "satellite"]
[options="nowrap" subs="+quotes,attributes"]
----
# cat /etc/rhsm/rhsm.conf
proxy_hostname = myproxy.example.com
proxy_port = 8080
proxy_user = optional_proxy_username
proxy_password = optional_proxy_password
----
endif::[]

ifeval::["{build}" != "satellite"]
----
echo "proxy=http://[2001:db8::::1]:8888" >> /etc/yum.conf
----
endif::[]

> == Hotfix for Foreman 1.24 or Katello 3.13

Reviewer 2: Since 2.1 is already GA, 1.24 is now unsupported. Safe to drop now.

NOTE: This should be removed for Foreman 2.0 or Satellite 6.8.

The following two installer patches are required to continue:

[options="nowrap" subs="+quotes,attributes"]
----
pushd /usr/share/foreman-installer/modules/dns
curl https://github.com/theforeman/puppet-dns/commit/7f5c514607f802e75081e50f1483888aca10d950.patch > dns-ip-fix.patch
patch -p1 -f < dns-ip-fix.patch
popd

pushd /usr/share/foreman-installer/modules/foreman_proxy
curl https://patch-diff.githubusercontent.com/raw/theforeman/puppet-foreman_proxy/pull/521.patch > dns-ip-fix.patch
patch -p1 -f < dns-ip-fix.patch
popd
----

Also one additional custom option must be explicitly set:

[options="nowrap" subs="+quotes,attributes"]
----
# # https://github.com/theforeman/puppet-foreman_proxy/pull/557
# cat >/etc/foreman-installer/custom-hiera.yaml <<EOF
foreman_proxy::plugin::dynflow::core_listen: "'::1'"
EOF

> ----

Author: Including this one I assume @ekohl ? Dropping thanks.

Reviewer 2: Yes, that is no longer needed.

== Configuring {Project} for UEFI HTTP Boot IPv6 provisioning

> To configure {Project} for UEFI HTTP Boot IPv6 provisioning, system checks must be disabled via `-s` and `--disable-system-checks` options, {SmartProxy} main HTTP endpoint must be enabled (the default port is 8000), TFTP and HTTP boot modules must be enabled as well. Discovery and DNS features can be optionally enabled however DHCP management must be either disabled in the installer or not used meaning that all IPv6 Subnets created in {Project} interface must have DHCP {SmartProxy} set to blank.

Reviewer 2: I'd prefer to use long options in documentation. So `--skip-checks-i-know-better` instead of `-s`.

However, I really want to understand why the checks should be disabled. Documentation should never recommend using those unless there's some BZ linked to track the fix.

Author: This will be removed, I don't remember why I had this.

Author: Hmmm was there a reason why this had to be there? I can't remember, damn. Maybe there is a reason.

Reviewer 3: @rdrazny checked installation for IPv6 today without the `--skip-checks` option and it worked fine.

[options="nowrap" subs="+quotes,attributes"]
----
foreman-installer \
 -v -s --no-colors \
 --disable-system-checks \

>  --scenario katello \

Reviewer 2: Why do we recommend the katello scenario?

Author: I am cleaning this up. Discovery is not needed, system check disable as well, password as well. It was just a copy paste from what I used during my testing.

Author: Cleaned up.

 --foreman-initial-admin-password changeme \
 --enable-foreman-cli-discovery \
 --enable-foreman-cli-templates \
 --enable-foreman-plugin-discovery \
 --enable-foreman-plugin-templates \
 --foreman-proxy-http true \
 --foreman-proxy-httpboot true \

>  --foreman-proxy-tftp true \

Author: Note to myself: add `--foreman-proxy-tftp-managed false` and info about this option (when to use it).

 --foreman-proxy-dhcp false \
 --foreman-proxy-dns true \
 --foreman-proxy-dns-zone "example.com" \

>  --foreman-proxy-dns-reverse "c.c.0.0.b.b.b.b.a.a.a.a.0.0.d.f.ip6.arpa" \

Author: Note for myself - the reverse zone is incorrect for this IPv6 example address.

 --enable-foreman-proxy-plugin-discovery \

>  --foreman-initial-admin-password changeme \

Reviewer 4: A duplicit option - this is already set at line 99

Author: Yeah I will clean it up a lot - most of these are not needed.

 --foreman-proxy-plugin-discovery-install-images true
 ----

NOTE: The following paragraph can be a module "How to set up HTTP proxy in {Project}" that can be reused across documents.

> If native IPv6 connectivity is not avilable, define HTTP proxy in {Project} and set it as the default proxy in settings.

Reviewer 2: ```suggestion
If native IPv6 connectivity is not available, define HTTP proxy in {Project} and set it as the default proxy in settings.
```

----
hammer http-proxy create --name "Default proxy" --url http://myproxy.example.com:8888 --organizations "Default Organization" --locations "Default Location"
hammer settings set --name content_default_http_proxy --value "Default proxy"
----

NOTE: The following can be a module/procedure. We need to drop a note that this is only relevant when Katello is installed if upstream.

Import a manifest and synchronize {RHEL} 8 kickstart repository.

== Creating new IPv6 Subnet

To create new IPv6 Subnet, go to Infrastructure - Subnets and create new entry:

Protocol: IPv6
Network Address: valid IPv6 address
Network Prefix: valid prefix
Primary DNS Server: must be valid IPv6 address of a DNS server

IPAM modes:

EUI-64 - IPv6 address is calculated from host provisioning interface MAC address
Internal DB - IPv6 address is selected from a sequence of all available addresses from the subnet
None - IPv6 address must be entered manually

Boot mode can be set both to DHCP or static, keep in mind that external DHCPv6 server is deployed and when this option is set to static a DHCPv6 service must be deployed on the network for the initial HTTP UEFI bootstrap.

> NOTE: Primary DNS server must be present becuase some older versions of {RHEL} 8 do not correctly interper DNS server coming from DHCPv6 response.

Reviewer 2: ```suggestion
NOTE: Primary DNS server must be present because some older versions of {RHEL} 8 do not correctly interpret DNS servers coming from DHCPv6 responses.
```

> To create a IPv6 subnet via hammer do:

Reviewer 2: ```suggestion
To create an IPv6 subnet via hammer do:
```

----
# hammer subnet create --name ipv6_subnet --network-type IPv6 --network 2001:db8:: --prefix 64 --tftp-id 1 --discovery-id 1 --httpboot-id 1 --template-id 1 --domains example.com
----

== Configuring Grub2 for HTTP UEFI

NOTE: This step is a workaround for the following bugs in grub2 component:
+
* https://bugzilla.redhat.com/show_bug.cgi?id=1732765
* https://bugzilla.redhat.com/show_bug.cgi?id=1763216
* https://bugzilla.redhat.com/show_bug.cgi?id=1616395

Replace `/var/lib/tftpboot/grub2/grubx64.efi` deployed by the {foreman-installer} with a version from Fedora stable release. The file can be copied from installed Fedora EFI system or extracted from `grub2-efi` RPM package:
+
[options="nowrap" subs="+quotes,attributes"]
----
# wget https://download.fedoraproject.org/pub/fedora/linux/releases/31/Server/x86_64/os/Packages/g/grub2-efi-x64-2.02-100.fc31.x86_64.rpm
# rpm2cpio grub2-efi-x64-2.02-100.fc31.x86_64.rpm | cpio -idmv
./boot/efi/EFI/fedora/fonts
./boot/efi/EFI/fedora/grubx64.efi
./boot/grub2/grubenv
./boot/loader/entries
./etc/grub2-efi.cfg
# cp ./boot/efi/EFI/fedora/grubx64.efi /var/lib/tftpboot/grub2/grubx64.efi
----

== Configuring external DHCPv6

> Both DHCPv6 and Router Advertisment services must be running on all provisioning networks. It can be installed on any host in the network, or on DHCP {SmartProxy}. {Project} currently does not manage DHCPv6 records, it is only used to bootstrap clients into installer.

Reviewer 2: ```suggestion
Both DHCPv6 and Router Advertisement services must be running on all provisioning networks. It can be installed on any host in the network, or on DHCP {SmartProxy}. {Project} currently does not manage DHCPv6 records, it is only used to bootstrap clients into the installer.
```

Author: Yeah thanks for these.

[options="nowrap" subs="+quotes,attributes"]
----
# cat /etc/dhcp/dhcpd6.conf
default-lease-time 2592000;
preferred-lifetime 604800;
option dhcp-renewal-time 3600;
option dhcp-rebinding-time 7200;
allow leasequery;
option dhcp6.info-refresh-time 21600;
option dhcp6.vendor-class-identifier code 16 = string;
option dhcp6.vendor-class code 16 = {integer 32, integer 16, string};
option dhcp6.user-class code 15 = string;
option dhcp6.bootfile-url code 59 = string;
option dhcp6.client-arch-type code 61 = array of unsigned integer 16;

subnet6 2001:db8::/64 {
  range6 2001:db8::aaaa 2001:db8::ffff;
  option dhcp6.name-servers 2001:db8::2;

  if exists dhcp6.user-class and
    substring(option dhcp6.user-class, 2, 4) = "iPXE" {
    # iPXE script file location
    option dhcp6.bootfile-url "http://{smartproxy-example-com}:8000/unattended/iPXE";
  } else if option dhcp6.client-arch-type = 00:06 {
    # UEFI x86 PXE/TFTP boot via IPv6

>     option dhcp6.bootfile-url "tftp://{smartproxy-example-com}:8000/ipxe32.efi";

Reviewer 2: If we don't manage the tftp server, this won't work, right? The current example does, but keeping https://github.com/theforeman/foreman-documentation/pull/116/files#r455082819 in mind.

Author: Yeah these are just examples, cases we don't care about.

    # UEFI x86 HTTP boot via IPv6
    option dhcp6.vendor-class 0 10 "HTTPClient";
    option dhcp6.bootfile-url "http://{smartproxy-example-com}:8000/EFI/grub2/grubx64.efi";
  } else if option dhcp6.client-arch-type = 00:10 {
    # UEFI x86-64 HTTP boot via IPv6
    option dhcp6.vendor-class 0 10 "HTTPClient";
    option dhcp6.bootfile-url "http://{smartproxy-example-com}:8000/EFI/grub2/grubx64.efi";
  } else {
    # Support a hypothetical BIOS system that can PXE boot over IPv6
    option dhcp6.bootfile-url "tftp://{smartproxy-example-com}:8000/undionly.kpxe";
  }
}
----

Example configuration for Router Advertisment server from {RHEL}:

[options="nowrap" subs="+quotes,attributes"]
----
# cat /etc/radvd.conf
interface eth0
{
        AdvSendAdvert on;
        AdvManagedFlag on;
        AdvOtherConfigFlag on;

        prefix 2001:db8::/64
        {
                AdvAutonomous off;
        };

        RDNSS 2001:db8::2 {
        };
};
----

NOTE: The DNS server must resolve the hostname in bootfile URL, `{smartproxy-example-com}` in the example above. It must also match DNS server set in the Subnet.

== Importing provisioning templates

> NOTE: This should be removed for Foreman 2.0 or Satellite 6.8.

Reviewer 2: Luckily these are now part of all supported releases :)

Author: Removed.

There are necessary template changes in PXE and Kickstart templates which do not ship with older versions of {Project}. To import them:

----
foreman-rake templates:import repo=https://github.com/lzap/community-templates branch=kickstart-ipv6-support prefix='' associate=1
----

== Building default template

Visit Hosts -> Templates and click on "Build PXE Default" to rebuild default template.

== Procedure: IPv6 UEFI HTTP Boot provisioning

NOTE: Watch the community demo! :-)

> https://community.theforeman.org/t/foreman-community-demo-73/16961

Reviewer 4: This link is wrong - the correct one is https://community.theforeman.org/t/foreman-community-demo-74/17201

Author: Removed anyway, this needs to be a procedure.
